packages/nx-aws-cache/src/tasks-runner/aws-cache.test.ts

```diff
@@ -0,0 +1,92 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import { randomUUID } from 'crypto';
+import * as path from 'path';
+import { mockClient } from 'aws-sdk-client-mock';
+import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
+import { Readable } from 'stream';
+import { AwsCache } from './aws-cache';
+import { Logger } from './logger';
+import { MessageReporter } from './message-reporter';
+
+// eslint-disable-next-line max-lines-per-function
+describe('Test aws put and get unencrypted file', () => {
+  let awsCache: AwsCache;
+  let uploadedData: Buffer | null;
+  const hash = randomUUID();
+  const cacheDirectory = path.join(os.tmpdir(), 'aws-cache');
+  const cacheDirectorySave = path.join(os.tmpdir(), 'aws-cache-decompress');
+  const fileContent = 'console.log(123)';
+  let filePath = '';
+
+  const config = {
+    encryptionFileKey: 'Pbfk58EpcK7IxTxWwSXNsTAKmzhJQE+99vkpGftyJg8=',
+    awsAccessKeyId: 'minio',
+    awsSecretAccessKey: 'minio123',
+    awsBucket: 'test',
+    awsEndpoint: 'http://127.0.0.1:9000',
+    awsForcePathStyle: true,
+    awsRegion: 'us-east-1',
+  };
+
+  beforeEach(() => {
+    uploadedData = null;
+    awsCache = new AwsCache(config, new MessageReporter(new Logger()));
+    const s3Mock = mockClient(S3Client);
+    s3Mock.on(GetObjectCommand).callsFake(() => {
+      if (!uploadedData) {
+        throw new Error('No upload data');
+      }
+      return {
+        Body: Readable.from(uploadedData),
+      };
+    });
+
+    // Aws-sdk-client-mock not mock CreateMultipartUploadCommand and UploadPartCommand for @aws-sdk/lib-storage I dont know why. There is a issue with this: https://github.com/m-radzikowski/aws-sdk-client-mock/issues/118
```

Comment:
I consider this a hack and I don't know the reason you did it like that. The same case is described in the docs: https://www.npmjs.com/package/aws-sdk-client-mock#s3-getobjectcommand

Here is the whole test spec which works fine:
```
import * as fs from 'fs';
import * as os from 'os';
import { randomUUID } from 'crypto';
import * as path from 'path';
import { mockClient } from 'aws-sdk-client-mock';
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { Readable } from 'stream';
import { AwsCache } from './aws-cache';
import { Logger } from './logger';
import { MessageReporter } from './message-reporter';
import { sdkStreamMixin } from '@aws-sdk/util-stream-node';
import { Encrypt, EncryptConfig } from './encryptor';

// eslint-disable-next-line max-lines-per-function
describe('Test aws put and get unencrypted file', () => {
  let uploadedData: Buffer | null;
  let awsCache: AwsCache;
  const s3Mock = mockClient(S3Client);
  const hash = randomUUID();
  const cacheDirectory = path.join(os.tmpdir(), 'aws-cache');
  const cacheDirectorySave = path.join(os.tmpdir(), 'aws-cache-decompress');
  const fileContent = 'console.log(123)';
  let filePath = '';

  const config = {
    encryptionFileKey: 'Pbfk58EpcK7IxTxWwSXNsTAKmzhJQE+99vkpGftyJg8=',
    awsAccessKeyId: 'minio',
    awsSecretAccessKey: 'minio123',
    awsBucket: 'test',
    awsEndpoint: 'http://127.0.0.1:9000',
    awsForcePathStyle: true,
    awsRegion: 'us-east-1',
  };

  beforeEach(() => {
    uploadedData = null;
    s3Mock.on(GetObjectCommand).callsFake(() => {
      if (!uploadedData) {
        throw new Error('No upload data');
      }
      return {
        Body: Readable.from(uploadedData),
      };
    });

    fs.mkdirSync(cacheDirectory, {
      recursive: true,
    });
    fs.mkdirSync(cacheDirectorySave, {
      recursive: true,
    });

    const fileDir = path.join(cacheDirectory, `${hash}/outputs`);

    fs.mkdirSync(fileDir, { recursive: true });
    filePath = path.join(fileDir, 'test.js');
    fs.writeFileSync(filePath, fileContent);
  });

  afterEach(() => {
    jest.resetAllMocks();
    s3Mock.reset();
  });

  it('Should save encrypted data in s3 file, and read an unencrypted', async () => {
    awsCache = new AwsCache(config, new MessageReporter(new Logger()));

    await awsCache.store(hash, cacheDirectory);

    const tgzFilePath = path.join(cacheDirectory, `${hash}.tar.gz`);
    const tgzFileStream = fs.createReadStream(tgzFilePath);
    const sdkStream = sdkStreamMixin(
      tgzFileStream.pipe(new Encrypt(new EncryptConfig(config.encryptionFileKey))),
    );
    s3Mock.on(GetObjectCommand).resolves({ Body: sdkStream });

    await awsCache.retrieve(hash, cacheDirectorySave);

    const extractedFilePath = path.join(cacheDirectorySave, `${hash}/outputs/test.js`);
    expect(fs.existsSync(extractedFilePath)).toBeTruthy();
    expect(fs.readFileSync(extractedFilePath).toString()).toBe(fileContent);
  });

  it('Should save in unencrypted s3 file, and read an unencrypted', async () => {
    const configWithoutEncryption = {
      ...config,
      encryptionFileKey: '',
    };
    awsCache = new AwsCache(configWithoutEncryption, new MessageReporter(new Logger()));

    await awsCache.store(hash, cacheDirectory);

    const tgzFilePath = path.join(cacheDirectory, `${hash}.tar.gz`);
    const tgzFileStream = fs.createReadStream(tgzFilePath);
    const sdkStream = sdkStreamMixin(tgzFileStream);

    s3Mock.on(GetObjectCommand).resolves({ Body: sdkStream });

    await awsCache.retrieve(hash, cacheDirectorySave);
    const extractedFilePath = path.join(cacheDirectorySave, `${hash}/outputs/test.js`);
    expect(fs.existsSync(extractedFilePath)).toBeTruthy();
    expect(fs.readFileSync(extractedFilePath).toString()).toBe(fileContent);
  });
});
```

Reply:
I've already pushed this code, didn't know I have rights to do it.

Reply:
Where in this code you store uploadedData ?

Reply:
Your tests is example of clean, fancy tests, but it worthless for tests end to end. 

I have created example of issue  on new branch test-error (https://github.com/gawsoftpl/nx-aws/tree/test-error) with
- your test: aws-cache.spec.ts
- my correct tests: aws-cache.e2e.spec.ts



For example if someone will try to simple PutObjectCommand not UploadMultipart your tests will be passed

### Example of Error code
https://github.com/gawsoftpl/nx-aws/blob/baaa2577e29524ad4e2de75c96c7a335e437a6bc/packages/nx-aws-cache/src/tasks-runner/aws-cache.ts#L185
```typescript
  private async uploadFile(hash: string, file: Readable): Promise<void> {
    const tgzFileName = this.getTgzFileName(hash);
    const params: clientS3.PutObjectCommand = new clientS3.PutObjectCommand({
      Bucket: this.bucket,
      Key: this.getS3Key(tgzFileName),
      Body: file,
    });

    try {
      this.logger.debug(`Storage Cache: Uploading ${hash}`);

      await this.s3.send(params);

      this.logger.debug(`Storage Cache: Stored ${hash}`);
    } catch (err) {
      throw new Error(`Storage Cache: Upload error - ${err}`);
    }
  }
```

### Correct code
https://github.com/gawsoftpl/nx-aws/blob/baaa2577e29524ad4e2de75c96c7a335e437a6bc/packages/nx-aws-cache/src/tasks-runner/aws-cache.ts#L211
```js
  private async uploadFile(hash: string, file: Readable) {
    try {
      this.logger.debug(`Storage Cache: Uploading ${hash}`);

      const tgzFileName = this.getTgzFileName(hash);

      const upload = new Upload({
        client: this.s3,
        params: {
          Bucket: this.bucket,
          Key: this.getS3Key(tgzFileName),
          Body: file,
        },
      });

      const response = await upload.done();
      this.logger.debug(`Storage Cache: Stored ${hash}`);

      return response;
    } catch (err) {
      throw new Error(`Storage Cache: Upload error - ${err}`);
    }
  }
```

I have created e2e tests with docker-compose: aws-cache.e2e.spec.ts

```sh
# Your tests will passed with Upload Multipart and PutObjectCommand
npx jest aws-cache.spec.ts

# My test will passed only with Upload Multipart, and failed on PutObjectCommand
aws-cache.e2e.spec.ts
```

### Why this is so important
Encrypt and Decrypt transform data with stream. To upload file with PutObjectCommand you HAVE to set contentLength before start upload. When you create stream transform you don't know what will be the final file size before the transformation is completed. 

When you create createReadStream you known file size on beginning, because you simple read file from disk. When you encrypt data you have to upload file with MultipartUpload.

Now your tests have issue because when I upload file to minio with PutObjectCommand. Minio will return success info without errors, logs etc, but minio will not save file in server, because you don't send final ContentLength. Upload process will be competed, but there will be no file in minio server.

That why I have created e2e tests with docker-compose, for protected code from change uploadfile to PutObjectCommand and for confirm that file is stored in server.

Your tests is only fancy Unit tests which test only that aws s3 command was executed, but you don't test that file in s3 is saved and encrypted. 

**With your test protection against hard-to-find bugs is only comment:**
'''When uploading a file with a transform stream, the final ContentLength is unknown so it has to be uploaded as multipart.'''

Run this
```sh
# Clone branch test-error from my fork
cd packages/nx-aws-cache
docker-compose up -d

# When you setup method uploadFile like this: https://github.com/gawsoftpl/nx-aws/blob/baaa2577e29524ad4e2de75c96c7a335e437a6bc/packages/nx-aws-cache/src/tasks-runner/aws-cache.ts#L185

# With PutObjectCommand file will be not saved in s3, tests failed
npx jest aws-cache.e2e.spec.ts

# With PutObjectCommand file will be not saved in s3, tests passed
npx jest aws-cache.spec.ts
```

Reply:
I've missed your comment and merged it before, you were late to comment. I understand the problem and yes, my example of tests is unit not e2e, it is a completely different thing. My 2 main reasons for commenting on this were:
- having e2e tests you should not rely on aws-cache.ts or encrypt functions directly, but test more from an e2e perspective (not really a user, but more a system with some input and output)
- using docker-compose for e2e is ok for local testing, but no for CI - you have to provide a way to run it with the same command on both environments and be consistent with nx targets

Having said this, feel free to open another PR where you can add that kind of test and I would be happy to review or suggest the solution.